test/ecto/query/inspect_test.exs

```diff
@@ -69,6 +69,35 @@ defmodule Ecto.Query.InspectTest do
            ~s{from p in subquery(from p in Inspect.Post)}
   end
 
+  test "CTE" do
+    initial_query =
+      "categories"
+      |> where([c], is_nil(c.parent_id))
+      |> select([c], %{id: c.id, depth: fragment("1")})
+
+    iteration_query =
+      "categories"
+      |> join(:inner, [c], t in "tree", on: t.id == c.parent_id)
+      |> select([c, t], %{id: c.id, depth: fragment("? + 1", t.depth)})
+
+    cte_query = initial_query |> union_all(^iteration_query)
+
+    query =
+      "products"
+      |> recursive_ctes(true)
+      |> with_cte("tree", as: ^cte_query)
+      |> join(:inner, [r], t in "tree", on: t.id == r.category_id)
+
+    assert i(query) ==
+      ~s{from p in "products", recursive_ctes: true, with_ctes: [tree: (from c in "categories",\n} <>
```

Comment:
This is tricky because we actually don't support this format.

Maybe we should emit this: `(from ...) |> with_cte(...) |> recursive_ctes(true)`?

Reply:
Ok. What about `#Ecto.Query<…>` form? Should it also be `(#Ecto.Query<…>) |> with_cte("name", as: #Ecto.Query<…>)`?

Reply:
Yes, that's what I would do, i think. No need to wrap the `#Ecto.Query<...>` one in parens though.